### PR TITLE
Upload instructions for mac and Rclone

### LINKED ISSUE #1973: Harvest: add rclone  example upload command on upload screens
Rclone is a great tool for transferring files - it supports checksums, retries, and progress stats.

Rclone uses different syntax - see https://rclone.org/docs/#backend-path-to-dir - for it's command line parameters which makes it a pain to use with our uploader.

- [ ] Add example rclone usage (maybe hidden) to upload screen prefilled with host/port/username/password to upload screens
- [ ] Maybe put it in another instructions tab? Windows/Max/Linux/rclone...

### PR DESCRIPTION
**Pull request by peichins**
# Upload instructions for mac and Rclone

## Changes

- Added instructions to the "mac" tab on the "uploading" stage of the harvest upload workflow.  This partly addresses #1971, but  the linux instructions are still missing.
- Added a new tab "Rclone" on that page with instructions for Rclone
  - this has very brief installation instructions and a commandline command the user can copy
  - The rclone command takes the host, port, username and password as separate arguments. The existing harvest class didn't store the port and host as separate properties. To access the port and host separately I created a new getter that returns the four values, using regex matching on the uploadUrl to extract the host and port. This resolves #1973 

## Issues

#1971, #1973

## Visual Changes

New tab:
image

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: src/app/components/harvest/screens/uploading/batch-uploading.component.html [hunk 1]
```diff
@@ -51,6 +76,40 @@ <h3>Uploading Files</h3>
...
+                --sftp-host {{ harvest.uploadArguments.host }} 
+                --sftp-port {{ harvest.uploadArguments.port }} 
+                "[path to your files]"
+                :sftp:/
+              </code>   
+            </li>
+
+            <li>
+              Note that rclone will copy the <em>contents</em> of the [path to your files] folder to the destination, not the source folder itself.
+              So if you have a folder for each site, specify their parent folder as the [path to your files] value.
```
Comment (atruskie) [inline src/app/components/harvest/screens/uploading/batch-uploading.component.html]: wrap `[path to your files]` in an inline code style

File: src/app/components/harvest/screens/uploading/batch-uploading.component.html [hunk 2]
```diff
@@ -51,6 +76,40 @@ <h3>Uploading Files</h3>
...
+            </li>
+
+            <li>
+              Note that rclone will copy the <em>contents</em> of the [path to your files] folder to the destination, not the source folder itself.
+              So if you have a folder for each site, specify their parent folder as the [path to your files] value. 
+            </li>
+          </ul>
+
+          
+
```
Comment (atruskie) [inline src/app/components/harvest/screens/uploading/batch-uploading.component.html]: empty lines?

File: src/app/models/Harvest.ts [hunk 3]
```diff
@@ -152,6 +159,20 @@ export class Harvest extends AbstractModel implements IHarvest {
...
     );
   }
+
+  public get uploadArguments(): HarvestUploadConnectionArguments {
+    return {
+      user: this.uploadUser,
+      // matches string of any character except colon after "sftp://"
+      host: this.uploadUrl.match(/sftp:\/\/([^:]+)/)[1],
+      // matches digits at the end after a colon
+      port: this.uploadUrl.match(/:([0-9]+)$/)[1],
```
Comment (atruskie) [inline src/app/models/Harvest.ts]: combine regexs and cache calculated values. These values could be added to the harvest model.

**Review by atruskie — CHANGES_REQUESTED** [review]
How was uploading with cyberduck?

- did the link correctly open the application?
- was the upload stable?
- did you complete a harvest on a mac?

Otherwise a few minor tweaks suggested. Good work.

File: src/app/components/harvest/screens/uploading/batch-uploading.component.html (hunk 1 above)
Comment (peichins) [reply src/app/components/harvest/screens/uploading/batch-uploading.component.html]: done

File: src/app/components/harvest/screens/uploading/batch-uploading.component.html (hunk 2 above)
Comment (peichins) [reply src/app/components/harvest/screens/uploading/batch-uploading.component.html]: gone now, this was an oversight

File: src/app/models/Harvest.ts (hunk 3 above)
Comment (peichins) [reply src/app/models/Harvest.ts]: done

**Comment by peichins** [comment]
Cyberduck does get launched when clicking the link and opens fine on mac.  I had no issues uploading 125 x 50Mb files.

File: src/app/components/harvest/screens/uploading/batch-uploading.component.html [hunk 4]
```diff
@@ -38,17 +38,97 @@ <h3>Uploading Files</h3>
...
+        <li>
+          On your first connection to our cloud, Cyberduck may prompt you with 
+          an "unknown fingerprint" message. Select "Allow".  
+        </li>
+        <li>
+          Upload all of your files by dragging them into the cyberduck window and we will check them for you. This  
+          <a target="_blank" href="https://docs.cyberduck.io/cyberduck/upload/"
+            >guide</a
+          >
+          has more help about uploading using cyberduck
```
Comment (atruskie) [inline src/app/components/harvest/screens/uploading/batch-uploading.component.html]:
```suggestion
          has more help about uploading using cyberduck.
```

File: src/app/components/harvest/screens/uploading/batch-uploading.component.html [hunk 5]
```diff
@@ -38,17 +38,97 @@ <h3>Uploading Files</h3>
...
+        <li>
+          Upload all of your files by dragging them into the cyberduck window and we will check them for you. This  
+          <a target="_blank" href="https://docs.cyberduck.io/cyberduck/upload/"
+            >guide</a
+          >
+          has more help about uploading using cyberduck.
+        </li>
+        <li>
+          Once done uploading, close Cyberduck before clicking on the Finished
+          Uploading button below
```
Comment (atruskie) [inline src/app/components/harvest/screens/uploading/batch-uploading.component.html]:
```suggestion
          Uploading button below.
```

File: src/app/components/harvest/screens/uploading/batch-uploading.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            spec.detectChanges();
+            if (!getConfirmCancelButton()) {
+                // if the click worked, the button will disappear and we can check the button action
+                expect(stages.transition).toHaveBeenCalledWith("complete");
+                clearInterval(interval);
+                done();
+            }
+        }, 150);
+
+    });
```
Comment (atruskie) [inline src/app/components/harvest/screens/uploading/batch-uploading.component.spec.ts]: We're going to need to do some work here calibrating effort. However some notes:

- why does the following test fail?
- maybe you could collapse the tests into one test, they're fairly similar after all
- I don't see tests for any of the other buttons? Now might be a good time to add them.

File: src/app/models/Harvest.ts [hunk 7]
```diff
@@ -132,6 +141,13 @@ export class Harvest extends AbstractModel implements IHarvest {
       (mapping) => new HarvestMapping(mapping, injector)
     );
     this.report = new HarvestReport(data.report, injector);
+    if (this.uploadUrl) {
+      this.uploadHost = this.uploadUrl.match(/sftp:\/\/([^:]+)/)[1];
+      this.uploadPort = parseInt(this.uploadUrl.match(/:([0-9]+)$/)[1], 10);
```
Comment (atruskie) [inline src/app/models/Harvest.ts]: there's no need to have two separate regexes here, combine them

File: src/app/models/Harvest.ts [hunk 8]
```diff
@@ -132,6 +141,13 @@ export class Harvest extends AbstractModel implements IHarvest {
...
+    }
+
+    //this.uploadHost = "";
+    //this.uploadPort = 2020;
```
Comment (atruskie) [inline src/app/models/Harvest.ts]:
```suggestion
```

**Review by atruskie — CHANGES_REQUESTED** [review]
Just a little more cleanup needed.

**Review by atruskie — APPROVED** [review]

**Comment by Allcharles** [comment]
FYI @atruskie, you should probably revoke my write access to the repo. Looks like i still have controls

**Pull request merged**